exonum/src/events/noise/wrappers/sodium_wrapper/wrapper.rs

```diff
@@ -123,55 +128,60 @@ impl NoiseWrapper {
     /// 3. Result message: first 4 bytes is message length(`len').
     /// 4. Append all encrypted packets in corresponding order.
     /// 5. Write result message to `buf`
-    pub fn encrypt_msg(&mut self, msg: &[u8], buf: &mut BytesMut) -> Result<Option<()>, io::Error> {
-        let len = self.encrypted_msg_len(msg.len());
-        let mut encrypted_message = Vec::with_capacity(len);
+    pub fn encrypt_msg(&mut self, msg: &[u8], buf: &mut BytesMut) -> io::Result<()> {
```

Comment:
FYI, I present a sketch of this method without additional allocations besides extending `buf`:

```rust
pub fn encrypt_msg(&mut self, msg: &[u8], buf: &mut BytesMut) -> io::Result<()> {
    const CHUNK_LENGTH: usize = MAX_MESSAGE_LENGTH - TAG_LENGTH;

    let message_len = encrypted_msg_len(msg.len()) + HEADER_LENGTH;
    buf.reserve(message_len);

    let start = buf.len();

    #[allow(unsafe_code)]
    unsafe {
        // We fill all bytes in the buffer in the cycle below.
        buf.set_len(start + message_len);
    }

    let encrypted_message = &mut buf[start..];

    LittleEndian::write_u32(
        &mut encrypted_message[..HEADER_LENGTH],
        (message_len - HEADER_LENGTH) as u32,
    );

    // Start of the last written chunk.
    let mut start = 0;
    // Length of the last written chunk.
    let mut len = HEADER_LENGTH;

    for (i, msg) in msg.chunks(CHUNK_LENGTH).enumerate() {
        debug_assert_eq!(start + len, HEADER_LENGTH + i * MAX_MESSAGE_LENGTH);
        start = HEADER_LENGTH + i * MAX_MESSAGE_LENGTH;
        len = self.write(msg, &mut encrypted_message[start..])?;
    }
    debug_assert_eq!(start + len, message_len);

    Ok(())
}
```

As it has unsafe code, it requires additional testing, so I'm against putting it in the code now. However, additional allocations clearly can be avoided.

Reply:
I like this idea to use only `buf` without additional allocations. Added TODO.